rust/blockstore/src/arrow/types.rs

```diff
@@ -18,13 +18,19 @@ pub trait ArrowWriteableKey: Key + Default {
 }
 
 pub trait ArrowWriteableValue: Value {
+    type ValueBuilder;
     type ReadableValue<'referred_data>: ArrowReadableValue<'referred_data>;
+    type PreparedValue;
 
     fn offset_size(item_count: usize) -> usize;
     fn validity_size(item_count: usize) -> usize;
     fn add(prefix: &str, key: KeyWrapper, value: Self, delta: &BlockDelta);
     fn delete(prefix: &str, key: KeyWrapper, delta: &BlockDelta);
     fn get_delta_builder() -> BlockStorage;
+    fn get_value_builder() -> Self::ValueBuilder;
+    fn prepare(value: Self) -> Self::PreparedValue;
```

Comment:
suggestions for alternatives to `prepare()` are welcome, some notes:

- may end up doing more work than necessary if `prepare()` is expensive and `set()` is called multiple times for the same key
- primitive types do not need `prepare()` and it may add overhead
- should `prepare()` also return a calculated size? calculating the size of most value types is cheap, some are expensive (e.g. DataRecord since it has to serialize a field into protobuf)

Currently, `prepare()` primarily simplifies implementation of `DataRecord` writing as the transient in-memory delta representation differs from the struct, but in the future it could also potentially improve efficiency by avoiding serializing multiple times to calculate the size.

Reply:
We could also sidestep this for now by making a separate trait, `ArrowWriteableValueColumn`, and put `get_value_builder()` / `append()` / `finish()` there.

Reply:
can we remove prepare()

Reply:
as a follow-up to our in-person discussion today: the reason we need this is that `ArrowWriteableValue::append()` for `&DataRecord<'_>` needs to take a `(String, Vec<f32>, Option<Vec<u8>>, Option<String>)` (owned) because that's what the builder stores, not a `&DataRecord`